libbeat/common/mapstr_test.go

```diff
@@ -802,6 +803,103 @@ func TestAddTagsWithKey(t *testing.T) {
 	}
 }
 
+func TestExpand(t *testing.T) {
+	type data struct {
+		Event       MapStr
+		Expected    MapStr
+		ErrExpected bool
+	}
+	tests := []data{
+		{
+			Event: MapStr{
+				"hello.world": 15,
+			},
+			Expected: MapStr{
+				"hello": MapStr{
+					"world": 15,
+				},
+			},
+		},
+		{
+			Event: MapStr{
+				"test": 15,
+			},
+			Expected: MapStr{
+				"test": 15,
+			},
+		},
+		{
+			Event: MapStr{
+				"test":           15,
+				"hello.there":    1,
+				"hello.world.ok": "test",
+				"elastic.for":    "search",
+			},
+			Expected: MapStr{
+				"test": 15,
+				"hello": MapStr{
+					"there": 1,
+					"world": MapStr{
+						"ok": "test",
+					},
+				},
+				"elastic": MapStr{
+					"for": "search",
+				},
+			},
+		},
+		{
+			Event: MapStr{
+				"root": MapStr{
+					"ok": 1,
+				},
+				"root.shared":        "yes",
+				"root.one.two.three": 4,
+			},
+			Expected: MapStr{
+				"root": MapStr{
+					"ok":     1,
+					"shared": "yes",
+					"one":    MapStr{"two": MapStr{"three": 4}},
+				},
+			},
+		},
+		{
+			Event: MapStr{
+				"over":     1,
+				"over.lap": 5,
+			},
+			ErrExpected: true,
+		},
+
+		// This is nondeterminstic
+		// 1. Put("root.one.two", 2)
+		// 2. Put("root.one", 1)
+		// yields: {"root":{"one":1}}
+		//
+		// The reverse causes an error: "expected map but type is int"
+		//
+		//{
+		//	Event: MapStr{
+		//		"root": MapStr{
+		//			"one": 1,
+		//		},
+		//		"root.one.two": 2,
+		//	},
+		//},
+	}
+
+	for _, test := range tests {
+		x, err := test.Event.Expand()
+		if test.ErrExpected {
+			require.Error(t, err)
+			continue
+		}
+		require.NoError(t, err)
+		assert.Equal(t, test.Expected, x)
+	}
```

Comment:
expand and flatten should fulfill the properties `flatten(expand(m)) == flatten(m)`, `expand(flatten(m)) == expand(m)`. I wonder if this is always the case.